File: database-driver-api/src/main/java/de/lino/database/export/archiv/ArchiveExporter.java
package de.lino.database.export.archiv;

import de.lino.database.export.data.DataExporter;
import de.lino.database.export.transcript.TranscriptExporter;

import java.io.IOException;
import java.nio.file.Path;

/**
 * The shape of a whole-directory, format-agnostic archive exporter, e.g.
 * {@link ExportCoordinator.DirectoryZipExporter}: writing everything under some source
 * location to a single file at {@code output}, as opposed to {@link DataExporter} and
 * {@link TranscriptExporter}, which each write one already-collected data set as a
 * table.
 *
 * <p>{@link ExportCoordinator.DirectoryZipExporter} implements this interface directly;
 * being a {@link FunctionalInterface}, it can just as well be satisfied with a lambda
 * or method reference for a one-off implementation.
 */
@FunctionalInterface
public interface ArchiveExporter {

    /**
     * Writes an archive to {@code output}.
     *
     * @param output the file path the archive is written to; overwritten if it already exists
     * @throws IOException if the archive cannot be written to {@code output}
     */
    void export(Path output) throws IOException;

}

File: python/database-driver-api/src/database_driver/api/utils/export/archiv/archive_exporter.py
"""Mirror of ``de.lino.database.utils.export.archiv.ArchiveExporter``."""

from __future__ import annotations

from pathlib import Path
from typing import Protocol, runtime_checkable


@runtime_checkable
class ArchiveExporter(Protocol):
    """The shape of a whole-directory, format-agnostic archive exporter, e.g. a
    ``DirectoryZipExporter``: writing everything under some source location to a single
    file at ``output``, as opposed to ``DataExporter`` and ``TranscriptExporter``, which
    each write one already-collected data set as a table.

    The Java original is a ``@FunctionalInterface``; the Python mirror is a structural
    ``Protocol`` for the same reason - any object with a matching ``export`` method
    satisfies it without inheriting, so a one-off implementation stays as lightweight as
    a Java lambda.
    """

    def export(self, output: Path) -> None:
        """Writes an archive to ``output``; overwritten if it already exists.

        Raises:
            OSError: If the archive cannot be written to ``output``.
        """
        ...
